File: src/main/java/com/fwdekker/randomness/string/StringSettingsDialog.java
package com.fwdekker.randomness.string;

import com.fwdekker.randomness.CapitalizationMode;
import com.fwdekker.randomness.JavaHelperKt;
import com.fwdekker.randomness.SettingsDialog;
import com.fwdekker.randomness.ui.ButtonGroupKt;
import com.fwdekker.randomness.ui.JIntSpinner;
import com.fwdekker.randomness.ui.JSpinnerRange;
import com.intellij.openapi.ui.ValidationInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.ButtonGroup;
import javax.swing.JComponent;
import javax.swing.JList;
import javax.swing.JPanel;
import javax.swing.ListSelectionModel;
import java.util.HashSet;


/**
 * Dialog for settings of random string generation.
 *
 * @see StringSettings
 * @see StringSettingsAction
 */
public final class StringSettingsDialog extends SettingsDialog<StringSettings> {
    private JPanel contentPane;
    private JSpinnerRange lengthRange;
    private JIntSpinner minLength;
    private JIntSpinner maxLength;
    private ButtonGroup enclosureGroup;
    private ButtonGroup capitalizationGroup;
    private JList<Alphabet> alphabetList;


    /**
     * Constructs a new {@code StringSettingsDialog} that uses the singleton {@code StringSettings} instance.
     */
    /* default */ StringSettingsDialog() {
        this(StringSettings.Companion.getDefault());
    }

    /**
     * Constructs a new {@code StringSettingsDialog} that uses the given {@code StringSettings} instance.
     *
     * @param settings the settings to manipulate with this dialog
     */
    /* default */ StringSettingsDialog(final @NotNull StringSettings settings) {
        super(settings);

        init();
        loadSettings();
    }


    @Override
    protected JComponent createCenterPanel() {
        return contentPane;
    }

    /**
     * Initialises custom UI components.
     * <p>
     * This method is called by the scene builder at the start of the constructor.
     */
    private void createUIComponents() {
        minLength = new JIntSpinner(1, 1);
        maxLength = new JIntSpinner(1, 1);
        lengthRange = new JSpinnerRange(minLength, maxLength, Integer.MAX_VALUE);

        alphabetList = new JList<>(Alphabet.values());
        alphabetList.setSelectionMode(ListSelectionModel.MULTIPLE_INTERVAL_SELECTION);
        alphabetList.setLayoutOrientation(JList.HORIZONTAL_WRAP);
    }


    @Override
    public void loadSettings(final @NotNull StringSettings settings) {
        minLength.setValue(settings.getMinLength());
        maxLength.setValue(settings.getMaxLength());
        ButtonGroupKt.setValue(enclosureGroup, settings.getEnclosure());
        ButtonGroupKt.setValue(capitalizationGroup, settings.getCapitalization());

        for (int i = 0; i < Alphabet.values().length; i++) {
            if (settings.getAlphabets().contains(Alphabet.values()[i])) {
                alphabetList.addSelectionInterval(i, i);
            }
        }
    }

    @Override
    public void saveSettings(final @NotNull StringSettings settings) {
        settings.setMinLength(minLength.getValue());
        settings.setMaxLength(maxLength.getValue());

        final String enclosure = ButtonGroupKt.getValue(enclosureGroup);
        settings.setEnclosure(enclosure == null ? StringSettings.DEFAULT_ENCLOSURE : enclosure);

        final String capitalizationMode = ButtonGroupKt.getValue(capitalizationGroup);
        settings.setCapitalization(capitalizationMode == null
            ? StringSettings.Companion.getDEFAULT_CAPITALIZATION()
            : CapitalizationMode.Companion.getMode(capitalizationMode));

        settings.setAlphabets(new HashSet<>(alphabetList.getSelectedValuesList()));
    }

    @Override
    @Nullable
    public ValidationInfo doValidate() {
        if (alphabetList.getSelectedValuesList().isEmpty())
            return new ValidationInfo("Please select at least one alphabet.", alphabetList);

        return JavaHelperKt.firstNonNull(
            minLength.validateValue(),
            maxLength.validateValue(),
            lengthRange.validateValue()
        );
    }
}

File: src/main/kotlin/com/fwdekker/randomness/string/StringSettingsDialog.kt
package com.fwdekker.randomness.string

import com.fwdekker.randomness.CapitalizationMode
import com.fwdekker.randomness.SettingsDialog
import com.fwdekker.randomness.ValidationException
import com.fwdekker.randomness.ui.ButtonGroupHelper
import com.fwdekker.randomness.ui.JLongSpinner
import com.fwdekker.randomness.ui.JSpinnerRange
import com.intellij.openapi.ui.ValidationInfo
import java.util.HashSet
import javax.swing.ButtonGroup
import javax.swing.JList
import javax.swing.JPanel
import javax.swing.ListSelectionModel


/**
 * Dialog for settings of random string generation.
 *
 * @param settings the settings to manipulate with this dialog. Defaults to [StringSettings.default]
 */
// TODO Kotlin-ify concatenation of alphabets
class StringSettingsDialog(settings: StringSettings = StringSettings.default) :
    SettingsDialog<StringSettings>(settings) {
    private lateinit var contentPane: JPanel
    private lateinit var lengthRange: JSpinnerRange
    private lateinit var minLength: JLongSpinner
    private lateinit var maxLength: JLongSpinner
    private lateinit var enclosureGroup: ButtonGroup
    private lateinit var capitalizationGroup: ButtonGroup
    private lateinit var alphabetList: JList<Alphabet>


    init {
        init()
        loadSettings()
    }


    override fun createCenterPanel() = contentPane

    /**
     * Initialises custom UI components.
     *
     * This method is called by the scene builder at the start of the constructor.
     */
    private fun createUIComponents() {
        minLength = JLongSpinner(1, 1, Integer.MAX_VALUE.toLong())
        maxLength = JLongSpinner(1, 1, Integer.MAX_VALUE.toLong())
        lengthRange = JSpinnerRange(minLength, maxLength, Integer.MAX_VALUE.toDouble())

        alphabetList = JList(Alphabet.values())
        alphabetList.selectionMode = ListSelectionModel.MULTIPLE_INTERVAL_SELECTION
        alphabetList.layoutOrientation = JList.HORIZONTAL_WRAP
    }


    override fun loadSettings(settings: StringSettings) {
        minLength.setValue(settings.minLength)
        maxLength.setValue(settings.maxLength)
        ButtonGroupHelper.setValue(enclosureGroup, settings.enclosure)
        ButtonGroupHelper.setValue(capitalizationGroup, settings.capitalization)

        for (i in 0 until Alphabet.values().size) {
            if (settings.alphabets.contains(Alphabet.values()[i])) {
                alphabetList.addSelectionInterval(i, i)
            }
        }
    }

    override fun saveSettings(settings: StringSettings) {
        settings.minLength = Math.toIntExact(minLength.value)
        settings.maxLength = Math.toIntExact(maxLength.value)
        settings.enclosure = ButtonGroupHelper.getValue(enclosureGroup)
        settings.capitalization = CapitalizationMode.getMode(ButtonGroupHelper.getValue(capitalizationGroup))
        settings.alphabets = HashSet(alphabetList.selectedValuesList)
    }

    override fun doValidate(): ValidationInfo? {
        try {
            minLength.validateValue()
            maxLength.validateValue()
            lengthRange.validate()
        } catch (e: ValidationException) {
            return ValidationInfo(e.message ?: "", e.component)
        }

        return if (alphabetList.selectedValuesList.isEmpty()) {
            ValidationInfo("Please select at least one option.", alphabetList)
        } else {
            null
        }
    }
}
